### PR TITLE
Remove "connected" attribute

### LINKED ISSUE #372: Rethink "connected" attribute
When we originally designed the ev3dev-lang libraries, we based them all on a concept where devices would expose a `connected` property to indicate whether they had successfully found a device matching the parameters from the constructor. This has led to a lot of code that looks like this:

```python
m = ev3.Motor(ev3.OUTPUT_A)
assert m.connected
```

...and lots of confused users as a result (especially when they don't know to include that assertion).

Do we want to change this behavior? I think it would make sense at this point to throw an exception from the constructor. This would be just for this library and wouldn't necessarily require a change to others.

### PR DESCRIPTION
**Pull request by WasabiFan**
Fixes #372

I haven't tested this on-device yet.

### CODE REVIEW

File: ev3dev/__init__.py [hunk 1]
```diff
@@ -124,13 +124,16 @@ def matches(attribute, pattern):
                 yield f
 
 
+class DeviceNotFoundException(Exception):
```
Comment (dwalton76) [inline ev3dev/__init__.py]: would do `class DeviceNotFound(Exception)`

File: ev3dev/__init__.py [hunk 2]
```diff
@@ -174,23 +177,24 @@ def get_index(file):
...
             try:
                 name = next(list_device_names(classpath, name_pattern, **kwargs))
                 self._path = classpath + '/' + name
                 self._device_index = get_index(name)
-                self.connected = True
             except StopIteration:
                 self._path = None
                 self._device_index = None
-                self.connected = False
+                raise DeviceNotFoundException("No device matching the given requirements was found.")
```
Comment (dwalton76) [inline ev3dev/__init__.py]: `"%s is not connected" % self` would let the user know exactly which device isn't connected

File: ev3dev/__init__.py [hunk 3]
```diff
@@ -256,7 +252,7 @@ def _raise_friendly_access_error(self, driver_error, attribute):
             # We will assume that a file-not-found error is the result of a disconnected device
             # rather than a library error. If that isn't the case, at a minimum the underlying
             # error info will be printed for debugging.
-            raise Exception("%s is no longer connected" % self) from driver_error
+            raise DeviceNotFoundException("%s is no longer connected" % self) from driver_error
```
Comment (dwalton76) [inline ev3dev/__init__.py]: s/DeviceNotFoundException/DeviceNotFound/

File: tests/api_tests.py [hunk 4]
```diff
@@ -21,25 +21,20 @@ def test_device(self):
...
-        self.assertTrue(not d.connected)
+        with self.assertRaises(ev3.DeviceNotFoundException):
+            d = ev3.Device('tacho-motor', 'motor*', address='outA', driver_name='not-valid')
 
         d = ev3.Device('lego-sensor', 'sensor*')
-        self.assertTrue(d.connected)
 
-        d = ev3.Device('this-does-not-exist')
-        self.assertFalse(d.connected)
+        with self.assertRaises(ev3.DeviceNotFoundException):
```
Comment (dwalton76) [inline tests/api_tests.py]: same for these two

Comment (dwalton76) [reply tests/api_tests.py]: other than that it looks good to me

Comment (WasabiFan) [reply tests/api_tests.py]: 😟 Why didn't our tests catch this? That's... quite worrying.

**Comment by dwalton76** [comment]
`I haven't tested this on-device yet.` @WasabiFan have you had a chance to test it on a device?  If so it looks good to me.

**Comment by WasabiFan** [comment]
@dwalton76 sorry, I'm quite busy right now so might not be able to take another look at these PRs for another few weeks. If I do get some time I'll report back.

**Comment by WasabiFan** [comment]
I got a chance to test this on my EV3. I instantiated a motor with motors connected and without and it operated as expected.

**Review by ddemidov — APPROVED** [review]

**Pull request merged**